src/Illuminate/Database/Eloquent/FactoryBuilder.php

```diff
@@ -131,7 +132,22 @@ protected function makeInstance(array $attributes = [])
 
             $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker, $attributes);
 
-            return new $this->class(array_merge($definition, $attributes));
+            $evaluated = $this->evaluateClosures(array_merge($definition, $attributes));
+
+            return new $this->class($evaluated);
         });
     }
+
+    /**
+     * Delay closures evaluation after merging to avoid duplication.
+     *
+     * @param  array  $attributes
+     * @return array
+     */
+    protected function evaluateClosures(array $attributes)
+    {
+        return array_map(function ($attribute) use ($attributes) {
+            return $attribute instanceof Closure ? $attribute($attributes) : $attribute;
```

Comment:
@GrahamCampbell Using `value` won't allow to pass the evaluated definitions as the closure attribute, which is essential to work with complex relationships:
```php
$factory->define(App\Post::class, function () {
    return [
        'user_id' => function () {
            return factory(App\User::class)->create()->id;
        },
        'user_type' => function (array $evaluated) {
            return App\User::find($evaluated['user_id'])->type;
        }
    ];
});
```